lib/coinbase/address/external_address.rb

```diff
@@ -46,10 +46,172 @@ def faucet
     end
 
+    # Builds a stake operation for the supplied asset. The stake operation
+    # may take a few minutes to complete in the case when infrastructure is spun up.
+    # @param amount [Integer,String,BigDecimal] The amount of the asset to stake
+    # @param asset_id [Symbol] The asset to stake
+    # @param mode [Symbol] The staking mode. Defaults to :default.
+    # @param options [Hash] Additional options for the stake operation
+    # @return [Coinbase::StakingOperation] The stake operation
+    def build_stake_operation(amount, asset_id, mode: :default, options: {})
+      validate_can_stake!(amount, asset_id, mode, options)
+
+      build_staking_operation(amount, asset_id, 'stake', mode: mode, options: options)
+    end
+
+    # Builds an unstake operation for the supplied asset.
+    # @param amount [Integer,String,BigDecimal] The amount of the asset to unstake
+    # @param asset_id [Symbol] The asset to unstake
+    # @param mode [Symbol] The staking mode. Defaults to :default.
+    # @param options [Hash] Additional options for the unstake operation
+    # @return [Coinbase::StakingOperation] The unstake operation
+    def build_unstake_operation(amount, asset_id, mode: :default, options: {})
+      validate_can_unstake!(amount, asset_id, mode, options)
+
+      build_staking_operation(amount, asset_id, 'unstake', mode: mode, options: options)
+    end
+
+    # Builds an claim_stake operation for the supplied asset.
+    # @param amount [Integer,String,BigDecimal] The amount of the asset to claim
+    # @param asset_id [Symbol] The asset to claim
+    # @param mode [Symbol] The staking mode. Defaults to :default.
+    # @param options [Hash] Additional options for the claim_stake operation
+    # @return [Coinbase::StakingOperation] The claim_stake operation
+    def build_claim_stake_operation(amount, asset_id, mode: :default, options: {})
+      validate_can_claim_stake!(amount, asset_id, mode, options)
+
+      build_staking_operation(amount, asset_id, 'claim_stake', mode: mode, options: options)
+    end
+
+    # Retreives the balances used for staking for the supplied asset.
+    # @param asset_id [Symbol] The asset to retrieve staking balances for
+    # @param mode [Symbol] The staking mode. Defaults to :default.
+    # @param options [Hash] Additional options for the staking operation
+    # @return [Hash] The staking balances
+    # @return [BigDecimal] :stakeable_balance The amount of the asset that can be staked
+    # @return [BigDecimal] :unstakeable_balance The amount of the asset that is currently staked and cannot be unstaked
+    # @return [BigDecimal] :claimable_balance The amount of the asset that can be claimed
+    def staking_balances(asset_id, mode: :default, options: {})
+      asset = Coinbase.call_api do
+        Coinbase::Asset.fetch(network_id, asset_id)
+      end
+
+      context_model = Coinbase.call_api do
+        stake_api.get_staking_context(
+          {
+            asset_id: asset_id,
```

Comment:
Are we handling `gwei` and `wei` server-side currently?

If the backend is expecting atomic units with the `eth` asset_id, then for `gwei` and `wei` we'll need to add special case client-side handling.

I'm planning on pushing that detail onto the server-side in an upcoming release.

For now though, we'll want to pass back the `primary_denomination` of the asset to the backend here:
https://github.com/coinbase/coinbase-sdk-ruby/blob/6fb94d565dbe45f2eb9eed3157f675bd9de28aaf/lib/coinbase/address/wallet_address.rb#L55

Then we should be getting back [Balance]( https://github.com/coinbase/coinbase-sdk-ruby/blob/v0.0.9/lib/coinbase/balance.rb) objects from the backend for these parameters and can leverage the same flow as:
https://github.com/coinbase/coinbase-sdk-ruby/blob/6fb94d565dbe45f2eb9eed3157f675bd9de28aaf/lib/coinbase/address/wallet_address.rb#L53-L60

Reply:
Looks like `build_staking_operation` is handling this properly!